File: questModule/src/main/java/ru/nekit/android/qls/quest/QuestMediatorFacade.java
package ru.nekit.android.qls.quest;

import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.BounceInterpolator;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.List;

import ru.nekit.android.qls.EventBus;
import ru.nekit.android.qls.R;
import ru.nekit.android.qls.quest.answer.common.IAnswerChecker;
import ru.nekit.android.qls.quest.common.Quest;
import ru.nekit.android.qls.quest.mediator.IQuestMediator;
import ru.nekit.android.qls.quest.mediator.answer.ButtonsQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.answer.IButtonsQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.answer.IQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.answer.ISpecialQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.content.EmptyQuestContentMediator;
import ru.nekit.android.qls.quest.mediator.content.IQuestContentMediator;
import ru.nekit.android.qls.quest.mediator.title.IQuestTitleMediator;
import ru.nekit.android.qls.utils.AnimationUtils;
import ru.nekit.android.qls.utils.KeyboardHost;
import ru.nekit.android.qls.utils.ViewHolder;

import static android.view.View.GONE;
import static android.view.View.INVISIBLE;
import static android.view.View.VISIBLE;
import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;
import static android.widget.RelativeLayout.ABOVE;
import static android.widget.RelativeLayout.BELOW;
import static ru.nekit.android.qls.lockScreen.content.LockScreenQuestViewContainerMediator.ACTION_SHOW_RIGHT_ANSWER_WINDOW;
import static ru.nekit.android.qls.lockScreen.content.LockScreenQuestViewContainerMediator.ACTION_SHOW_WRONG_ANSWER_WINDOW;
import static ru.nekit.android.qls.quest.QuestContext.QuestState.DELAYED_PLAY;
import static ru.nekit.android.qls.quest.QuestContext.QuestState.PLAYED;
import static ru.nekit.android.qls.quest.QuestContextEvent.EVENT_QUEST_ATTACH;
import static ru.nekit.android.qls.quest.QuestContextEvent.EVENT_QUEST_PAUSE;
import static ru.nekit.android.qls.quest.QuestContextEvent.EVENT_QUEST_PLAY;
import static ru.nekit.android.qls.quest.QuestContextEvent.EVENT_QUEST_REPLAY;
import static ru.nekit.android.qls.quest.QuestContextEvent.EVENT_QUEST_RESUME;
import static ru.nekit.android.qls.quest.QuestContextEvent.EVENT_QUEST_START;
import static ru.nekit.android.qls.quest.QuestContextEvent.EVENT_QUEST_STOP;
import static ru.nekit.android.qls.quest.QuestContextEvent.EVENT_RIGHT_ANSWER;
import static ru.nekit.android.qls.quest.QuestContextEvent.EVENT_WRONG_ANSWER;

public class QuestMediatorFacade implements View.OnClickListener, IQuestMediatorFacade,
        EventBus.IEventHandler,
        View.OnLayoutChangeListener {

    private QuestContext mQuestContext;
    private Quest mQuest;
    private QuestViewHolder mViewHolder;
    @NonNull
    private IQuestTitleMediator mTitleMediator;
    @NonNull
    private IQuestContentMediator mContentMediator;
    @NonNull
    private IQuestAnswerMediator mAnswerMediator;
    @NonNull
    private IAnswerChecker mAnswerChecker;

    QuestMediatorFacade(@NonNull QuestContext questContext,
                        @NonNull IAnswerChecker answerChecker,
                        @NonNull IQuestTitleMediator questTitleMediator,
                        @Nullable IQuestContentMediator questContentMediator,
                        @Nullable IQuestAnswerMediator questAnswerMediator) {
        mQuestContext = questContext;
        mAnswerChecker = answerChecker;
        mTitleMediator = questTitleMediator;
        mContentMediator = questContentMediator == null ?
                new EmptyQuestContentMediator() : questContentMediator;
        mAnswerMediator = questAnswerMediator == null ?
                new ButtonsQuestAnswerMediator() : questAnswerMediator;
        questContext.getEventBus().handleEvents(this,
                EVENT_QUEST_ATTACH,
                EVENT_QUEST_START,
                EVENT_QUEST_PLAY,
                EVENT_QUEST_REPLAY,
                EVENT_QUEST_PAUSE,
                EVENT_QUEST_RESUME,
                EVENT_QUEST_STOP,
                EVENT_WRONG_ANSWER,
                EVENT_RIGHT_ANSWER
        );
    }

    @Override
    public void updateSize() {
        /*boolean contentViewIsPresent = false, alternativeAnswerIsPresent = false;
        if (mQuestContext.questHasState(PLAYED) && !mQuestContext.questHasState(PAUSED)) {
            contentViewIsPresent = mContentMediator.getView() != null;
            alternativeAnswerIsPresent =
                    mViewHolder.alternativeAnswerContainer.getVisibility() == VISIBLE;
            if (contentViewIsPresent) {
                mContentMediator.updateSize();
            }
            if (alternativeAnswerIsPresent) {
                mAnswerMediator.updateSize();
            }
        }
        if (contentViewIsPresent) {
            mViewHolder.contentContainer.requestLayout();
        }
        if (alternativeAnswerIsPresent) {
            mViewHolder.alternativeAnswerContainer.requestLayout();
        }*/
    }

    private boolean questDelayedPlay() {
        return mQuestContext.questHasState(DELAYED_PLAY);
    }

    @Override
    public void onCreate(@NonNull QuestContext questContext) {
        mQuestContext = questContext;
        mQuest = questContext.getQuest();
        mViewHolder = new QuestViewHolder(questContext);
        mTitleMediator.onCreate(questContext);
        mContentMediator.onCreate(questContext);
        mAnswerMediator.onCreate(questContext);
        mViewHolder.titleContainer.addView(mTitleMediator.getView());
        View contentView = mContentMediator.getView();
        RelativeLayout.LayoutParams answerContainerLayoutParams =
                (RelativeLayout.LayoutParams) mViewHolder.answerContainer.getLayoutParams();
        ViewGroup.LayoutParams answerButtonContainerLayoutParams =
                mViewHolder.alternativeAnswerContainer.getLayoutParams();
        if (mContentMediator.getView() == null) {
            answerContainerLayoutParams.addRule(BELOW, R.id.container_title);
            answerButtonContainerLayoutParams.height = MATCH_PARENT;
        } else {
            answerContainerLayoutParams.addRule(BELOW, 0);
            answerButtonContainerLayoutParams.height = WRAP_CONTENT;
        }
        RelativeLayout.LayoutParams contentContainerLayoutParams =
                (RelativeLayout.LayoutParams) mViewHolder.contentContainer.getLayoutParams();
        if (mContentMediator.includeInLayout()) {
            contentContainerLayoutParams.addRule(ABOVE, R.id.container_answer);
            contentContainerLayoutParams.addRule(BELOW, R.id.container_title);
        } else {
            contentContainerLayoutParams.setMargins(0, 0, 0, 0);
            contentContainerLayoutParams.addRule(ABOVE, 0);
            contentContainerLayoutParams.addRule(BELOW, 0);
        }
        if (contentView != null) {
            mViewHolder.contentContainer.addView(mContentMediator.getView());
        }
        mAnswerMediator.setAnswerCallback(questContext);
        mAnswerMediator.setAnswerChecker(mAnswerChecker);
        View answerView = mAnswerMediator.getView();
        if (answerView == null) {
            if (mAnswerMediator instanceof IButtonsQuestAnswerMediator) {
                IButtonsQuestAnswerMediator answerMediator =
                        (IButtonsQuestAnswerMediator) mAnswerMediator;
                List<View> buttonList = answerMediator.getAnswerButtonList();
                if (buttonList != null) {
                    for (View button : buttonList) {
                        mViewHolder.alternativeAnswerContainer.addView(button);
                    }
                }
            }
        } else {
            mViewHolder.alternativeAnswerContainer.addView(answerView);
        }
        boolean defaultAnswerInputVisible = !alternativeAnswerIsPresent();
        mViewHolder.defaultAnswerInputContainer.setVisibility(defaultAnswerInputVisible ? VISIBLE : GONE);
        if (defaultAnswerInputVisible) {
            mViewHolder.deaultAnswerInput.setVisibility(defaultAnswerInputVisible ? VISIBLE : GONE);
            boolean showAnswerButton = !answerButtonIsPresent() && questContext.answerButtonVisible();
            mViewHolder.deaultAnswerButton.setVisibility(showAnswerButton ? VISIBLE : GONE);
            if (!defaultAnswerInputVisible && showAnswerButton) {
                LinearLayout.LayoutParams answerButtonLayoutParameters =
                        (LinearLayout.LayoutParams) mViewHolder.deaultAnswerButton.getLayoutParams();
                answerButtonLayoutParameters.weight = 1;
                mViewHolder.deaultAnswerButton.setLayoutParams(answerButtonLayoutParameters);
            }
            mViewHolder.alternativeAnswerContainer.setVisibility(answerButtonIsPresent() ? VISIBLE : GONE);
            if (mViewHolder.deaultAnswerButton.getVisibility() == VISIBLE) {
                mViewHolder.deaultAnswerButton.setOnClickListener(this);
            }
            EditText answerInput = null;
            if (answerInputIsPresent()) {
                EditText contentAnswerInput = mContentMediator.getAnswerInput();
                if (contentAnswerInput != null && contentAnswerInput.getVisibility() == VISIBLE) {
                    contentAnswerInput.setInputType(mQuest.getAnswerInputType());
                    answerInput = contentAnswerInput;
                }
            } else {
                answerInput = mViewHolder.deaultAnswerInput;
            }
            if (answerInput != null) {
                answerInput.setInputType(mQuest.getAnswerInputType());
                answerInput.setImeOptions(EditorInfo.IME_ACTION_DONE);
                answerInput.setOnEditorActionListener(
                        new EditText.OnEditorActionListener() {
                            @Override
                            public boolean onEditorAction(TextView textView, int actionId, KeyEvent event) {
                                if (actionId == EditorInfo.IME_ACTION_DONE) {
                                    answerFunction();
                                    return true;
                                }
                                return false;
                            }
                        });
            }
        }
        updateVisibilityOfViews(!questDelayedPlay());
    }

    @Override
    public void onQuestAttach(@NonNull ViewGroup rootContentContainer) {
        mTitleMediator.onQuestAttach(mViewHolder.titleContainer);
        mContentMediator.onQuestAttach(mViewHolder.contentContainer);
        mAnswerMediator.onQuestAttach(mViewHolder.alternativeAnswerContainer);
        mViewHolder.getView().addOnLayoutChangeListener(this);
    }

    @Override
    public void onQuestStart(boolean delayedPlay) {
        mViewHolder.titleContainer.requestLayout();
        mViewHolder.contentContainer.requestLayout();
        mViewHolder.alternativeAnswerContainer.requestLayout();
        mTitleMediator.onQuestStart(delayedPlay);
        mContentMediator.onQuestStart(delayedPlay);
        mAnswerMediator.onQuestStart(delayedPlay);
        requestFocus();
    }

    @Override
    public void onQuestPlay(boolean delayedPlay) {
        updateVisibilityOfViews(true);
        mTitleMediator.onQuestPlay(delayedPlay);
        mContentMediator.onQuestPlay(delayedPlay);
        if (delayedPlay) {
            if (answerButtonIsPresent()) {
                mAnswerMediator.onQuestPlay(true);
            } else {
                int duration = mQuestContext.getQuestDelayedPlayAnimationDuration();
                View view = mViewHolder.answerContainer;
                if (view != null) {
                    view.setAlpha(0);
                    view.animate().withLayer().alpha(1).setDuration(duration);
                }
            }
        } else {
            mAnswerMediator.onQuestPlay(false);
        }
        requestFocus();
    }

    @Override
    public boolean onRightAnswer() {
        boolean showRightAnswerWindow = true;
        showRightAnswerWindow = showRightAnswerWindow && mTitleMediator.onRightAnswer();
        showRightAnswerWindow = showRightAnswerWindow && mContentMediator.onRightAnswer();
        showRightAnswerWindow = showRightAnswerWindow && mAnswerMediator.onRightAnswer();
        return showRightAnswerWindow;
    }

    @Override
    public boolean onWrongAnswer() {
        boolean showWrongAnswerWindow = true;
        showWrongAnswerWindow = showWrongAnswerWindow && mTitleMediator.onWrongAnswer();
        showWrongAnswerWindow = showWrongAnswerWindow && mContentMediator.onWrongAnswer();
        showWrongAnswerWindow = showWrongAnswerWindow && mAnswerMediator.onWrongAnswer();
        return showWrongAnswerWindow;
    }

    @Override
    public void onQuestReplay() {
        mTitleMediator.onQuestReplay();
        mContentMediator.onQuestReplay();
        mAnswerMediator.onQuestReplay();
        requestFocus();
    }

    @Override
    public void onQuestPause() {
        mTitleMediator.onQuestPause();
        mContentMediator.onQuestPause();
        mAnswerMediator.onQuestPause();
    }

    @Override
    public void onQuestResume() {
        mTitleMediator.onQuestResume();
        mContentMediator.onQuestResume();
        mAnswerMediator.onQuestResume();
        requestFocus();
    }

    @Override
    public void onQuestStop() {
        mTitleMediator.onQuestStop();
        mContentMediator.onQuestStop();
        mAnswerMediator.onQuestStop();
    }

    private boolean answerInputIsPresent() {
        EditText answerInput = mContentMediator.getAnswerInput();
        return answerInput != null && answerInput.getVisibility() == VISIBLE;
    }

    private boolean answerButtonIsPresent() {
        if (mAnswerMediator instanceof IButtonsQuestAnswerMediator) {
            IButtonsQuestAnswerMediator answerMediator =
                    (IButtonsQuestAnswerMediator) mAnswerMediator;
            List<View> buttonList = answerMediator.getAnswerButtonList();
            View answerView = mAnswerMediator.getView();
            return (buttonList != null && buttonList.size() > 0) ||
                    (answerView != null && answerView.getVisibility() == VISIBLE);
        }
        return false;
    }

    private boolean alternativeAnswerIsPresent() {
        return answerButtonIsPresent()
                || answerInputIsPresent()
                || mAnswerMediator instanceof ISpecialQuestAnswerMediator;
    }

    @Override
    public void onClick(View view) {
        answerFunction();
    }

    private void answerFunction() {
        if (!answerButtonIsPresent()) {
            CharSequence answerCharSequence = (answerInputIsPresent() ?
                    mContentMediator.getAnswerInput() : mViewHolder.deaultAnswerInput).getText();
            if (answerCharSequence != null && answerCharSequence.length() > 0) {
                String answerString = answerCharSequence.toString();
                if (mAnswerChecker.checkStringInputFormat(mQuest, answerString)) {
                    if (mAnswerChecker.checkStringInput(mQuest, answerString)) {
                        mQuestContext.rightAnswer();
                    } else {
                        mQuestContext.wrongAnswer();
                    }
                } else {
                    mQuestContext.wrongStringInputFormat();
                    setErrorColorAndReturnToNormal();
                }
            } else {
                mQuestContext.emptyAnswer();
                setErrorColorAndReturnToNormal();
            }
        }
    }

    private void setErrorColorAndReturnToNormal() {
        AnimationUtils.getColorAnimator(mQuestContext, R.color.green,
                R.color.red,
                mQuestContext.getResources().getInteger(R.integer.short_animation_duration),
                mViewHolder.defaultAnswerInputContainer, new BounceInterpolator()
        ).start();
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                AnimationUtils.getColorAnimator(mQuestContext, R.color.red,
                        R.color.green,
                        mQuestContext.getResources().getInteger(R.integer.short_animation_duration),
                        mViewHolder.defaultAnswerInputContainer
                ).start();
            }
        }, 1000);
    }

    public View getView() {
        return mViewHolder.getView();
    }

    @Override
    public void deactivate() {
        getView().removeOnLayoutChangeListener(this);
        mViewHolder.deaultAnswerInput.setOnEditorActionListener(null);
        mQuestContext.getEventBus().stopHandleEvents(this);
        onQuestStop();
        mTitleMediator.deactivate();
        mContentMediator.deactivate();
        mAnswerMediator.deactivate();
    }

    @Override
    public void detachView() {
        mTitleMediator.detachView();
        mContentMediator.detachView();
        mAnswerMediator.detachView();
    }

    @Override
    public IQuestMediator getTitleMediator() {
        return mTitleMediator;
    }

    @Override
    public IQuestMediator getContentMediator() {
        return mContentMediator;
    }

    @Override
    public IQuestMediator getAlternativeAnswerMediator() {
        return mAnswerMediator;
    }

    private void requestFocus() {
        if (questDelayedPlay() && !mQuestContext.questHasState(PLAYED)) {
            KeyboardHost.hideKeyboard(mQuestContext, mViewHolder.deaultAnswerInput, null);
        } else {
            if (answerButtonIsPresent()) {
                KeyboardHost.hideKeyboard(mQuestContext, mViewHolder.deaultAnswerInput);
            }
            if (mContentMediator.getAnswerInput() == null) {
                if (mViewHolder.deaultAnswerInput.getVisibility() == VISIBLE) {
                    KeyboardHost.showKeyboard(mQuestContext, mViewHolder.deaultAnswerInput);
                } else {
                    KeyboardHost.hideKeyboard(mQuestContext, mViewHolder.deaultAnswerInput);
                }
            } else {
                if (mContentMediator.getAnswerInput().getVisibility() == VISIBLE) {
                    KeyboardHost.showKeyboard(mQuestContext, mContentMediator.getAnswerInput());
                }
            }
        }
    }

    @Override
    public void onEvent(@NonNull Intent intent) {
        String action = intent.getAction();
        switch (action) {

            case EVENT_QUEST_ATTACH:

                onQuestAttach(mViewHolder.rootContainer);

                break;

            case EVENT_QUEST_START:

                onQuestStart(questDelayedPlay());

                break;

            case EVENT_QUEST_PLAY:

                onQuestPlay(questDelayedPlay());

                break;

            case EVENT_QUEST_REPLAY:

                onQuestReplay();

                break;

            case EVENT_QUEST_PAUSE:

                onQuestPause();

                break;

            case EVENT_QUEST_RESUME:

                onQuestResume();

                break;

            case EVENT_QUEST_STOP:

                onQuestStop();

                break;

            case EVENT_RIGHT_ANSWER:

                if (onRightAnswer()) {
                    mQuestContext.getEventBus().sendEvent(ACTION_SHOW_RIGHT_ANSWER_WINDOW);
                }

                break;

            case EVENT_WRONG_ANSWER:

                if (onWrongAnswer()) {
                    mQuestContext.getEventBus().sendEvent(ACTION_SHOW_WRONG_ANSWER_WINDOW);
                }

                break;

        }
    }

    private void updateVisibilityOfViews(boolean value) {
        View contentView = mContentMediator.getView();
        if (contentView == null) {
            mViewHolder.contentContainer.setVisibility(View.GONE);
        } else {
            mViewHolder.contentContainer.setVisibility(value ? VISIBLE : INVISIBLE);
        }
        if (answerButtonIsPresent()) {
            mViewHolder.alternativeAnswerContainer.setVisibility(value ? VISIBLE : INVISIBLE);
        } else {
            mViewHolder.answerContainer.setVisibility(value ? VISIBLE : INVISIBLE);
        }
    }

    @NonNull
    @Override
    public String getEventBusName() {
        return getClass().getName();
    }

    @Override
    public void onLayoutChange(View view, int left, int top, int right, int bottom, int oldLeft,
                               int oldTop, int oldRight, int oldBottom) {
        updateSize();
    }

    static class QuestViewHolder extends ViewHolder {

        @NonNull
        final ViewGroup rootContainer, titleContainer, answerContainer, contentContainer,
                alternativeAnswerContainer, defaultAnswerInputContainer;
        @NonNull
        final View deaultAnswerButton;
        @NonNull
        final EditText deaultAnswerInput;

        QuestViewHolder(@NonNull Context context) {
            super(context, R.layout.layout_quest_view);
            rootContainer = (ViewGroup) mView.findViewById(R.id.container_root);
            titleContainer = (ViewGroup) mView.findViewById(R.id.container_title);
            answerContainer = (ViewGroup) mView.findViewById(R.id.container_answer);
            contentContainer = (ViewGroup) mView.findViewById(R.id.container_content);
            alternativeAnswerContainer = (ViewGroup) mView.findViewById(R.id.container_answer_alternative);
            defaultAnswerInputContainer = (ViewGroup) mView.findViewById(R.id.container_answer_default);
            deaultAnswerButton = mView.findViewById(R.id.btn_answer_default);
            deaultAnswerInput = (EditText) mView.findViewById(R.id.input_answer_default);
        }
    }
}

File: questPresentation/src/main/java/ru/nekit/android/qls/quest/view/QuestMediatorFacade.kt
package ru.nekit.android.qls.quest.view

import android.content.Context
import android.view.View
import android.view.View.*
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import android.view.animation.BounceInterpolator
import android.view.inputmethod.EditorInfo.IME_ACTION_DONE
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.RelativeLayout.ABOVE
import android.widget.RelativeLayout.BELOW
import com.jakewharton.rxbinding2.widget.editorActionEvents
import io.reactivex.Completable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.Subject
import ru.nekit.android.domain.event.KeyboardHideAction
import ru.nekit.android.qls.R
import ru.nekit.android.qls.data.representation.getAnswerInputType
import ru.nekit.android.qls.domain.model.AnswerType
import ru.nekit.android.qls.domain.model.QuestState.DELAYED_PLAY
import ru.nekit.android.qls.domain.model.QuestState.PLAYED
import ru.nekit.android.qls.domain.model.quest.Quest
import ru.nekit.android.qls.quest.QuestContext
import ru.nekit.android.qls.quest.QuestContextEvent
import ru.nekit.android.qls.quest.QuestContextEvent.*
import ru.nekit.android.qls.quest.view.mediator.answer.ButtonListAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.answer.IAlternativeAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.answer.IAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.answer.IButtonListAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.content.EmptyQuestContentMediator
import ru.nekit.android.qls.quest.view.mediator.content.IContentMediator
import ru.nekit.android.qls.quest.view.mediator.title.ITitleMediator
import ru.nekit.android.qls.utils.KeyboardHost
import ru.nekit.android.qls.utils.KeyboardHost.hideKeyboard
import ru.nekit.android.qls.utils.throttleClicks
import ru.nekit.android.utils.AnimationUtils
import ru.nekit.android.utils.Delay
import ru.nekit.android.utils.ViewHolder
import java.util.concurrent.TimeUnit

//ver 1.2
class QuestMediatorFacade internal constructor(override var questContext: QuestContext,
                                               override val titleMediator: ITitleMediator,
                                               contentMediator: IContentMediator?,
                                               questAnswerMediator: IAnswerMediator?) :
        IQuestMediatorFacade,
        View.OnLayoutChangeListener {

    override var disposable: CompositeDisposable = CompositeDisposable()

    override val view: View
        get() = viewHolder.view

    override val contentMediator: IContentMediator = contentMediator
            ?: EmptyQuestContentMediator()
    override val answerMediator: IAnswerMediator = questAnswerMediator
            ?: ButtonListAnswerMediator()

    private lateinit var viewHolder: QuestViewHolder

    override lateinit var quest: Quest

    private val answerCallback: Subject<Any>
        get() = questContext.answerCallback

    init {
        listenForEvent(KeyboardHideAction::class.java) { action ->
            hideKeyboard(questContext, viewHolder.defaultAnswerInput, Delay.KEYBOARD.get(questContext), action.action)
        }
        listenForEvent(QuestContextEvent::class.java) {
            when (it) {
                QUEST_ATTACH -> onQuestAttach(viewHolder.rootContainer)
                QUEST_START -> questHasState(DELAYED_PLAY) {
                    onQuestStart(it)
                }
                QUEST_PLAY -> questHasState(DELAYED_PLAY) {
                    onQuestPlay(it)
                }
                QUEST_REPLAY -> onQuestReplay()
                QUEST_PAUSE -> onQuestPause()
                QUEST_RESUME -> onQuestResume()
                QUEST_STOP -> onQuestStop()
                EMPTY_ANSWER -> {
                    onAnswer(AnswerType.EMPTY)
                    setErrorColorAndReturnToNormal()
                }
                WRONG_INPUT_FORMAT_ANSWER -> {
                    onAnswer(AnswerType.WRONG_INPUT_FORMAT)
                    setErrorColorAndReturnToNormal()
                }
                else -> {
                }
            }
        }
    }

    override fun updateSize() {
        /*boolean contentViewIsPresent = false, alternativeAnswerIsPresent = false;
        if (questContext.questHasState(PLAYED) && !questContext.questHasState(PAUSED)) {
            contentViewIsPresent = contentMediator.getView() != null;
            alternativeAnswerIsPresent =
                    viewHolder.alternativeAnswerContainer.getVisibility() == VISIBLE;
            if (contentViewIsPresent) {
                contentMediator.updateSize();
            }
            if (alternativeAnswerIsPresent) {
                answerMediator.updateSize();
            }
        }
        if (contentViewIsPresent) {
            viewHolder.contentContainer.requestLayout();
        }
        if (alternativeAnswerIsPresent) {
            viewHolder.alternativeAnswerContainer.requestLayout();
        }*/
    }

    private fun answerButtonVisible(): Boolean {
        val ta = questContext.obtainStyledAttributes(R.style.Quest, R.styleable.QuestStyle)
        val showAnswerButton = ta.getBoolean(R.styleable.QuestStyle_answerButtonVisible, false)
        ta.recycle()
        return showAnswerButton
    }

    override fun onCreate(questContext: QuestContext, quest: Quest) {
        this.questContext = questContext
        this.quest = quest
        viewHolder = QuestViewHolder(questContext)
        answerMediator.answerPublisher = answerCallback
        titleMediator.onCreate(questContext, quest)
        contentMediator.onCreate(questContext, quest)
        answerMediator.onCreate(questContext, quest)
        viewHolder.titleContainer.addView(titleMediator.view)
        val contentView = contentMediator.view
        val answerContainerLayoutParams = viewHolder.answerContainer.layoutParams as RelativeLayout.LayoutParams
        val answerButtonContainerLayoutParams = viewHolder.alternativeAnswerContainer.layoutParams
        if (contentView == null) {
            answerContainerLayoutParams.addRule(BELOW, R.id.container_title)
            answerButtonContainerLayoutParams.height = MATCH_PARENT
        } else {
            answerContainerLayoutParams.addRule(BELOW, 0)
            answerButtonContainerLayoutParams.height = WRAP_CONTENT
        }
        (viewHolder.contentContainer.layoutParams as RelativeLayout.LayoutParams).apply {
            if (contentMediator.includeInLayout()) {
                addRule(ABOVE, R.id.container_answer)
                addRule(BELOW, R.id.container_title)
            } else {
                setMargins(0, 0, 0, 0)
                addRule(ABOVE, 0)
                addRule(BELOW, 0)
            }
        }
        if (contentView != null) {
            viewHolder.contentContainer.addView(contentMediator.view)
        }
        viewHolder.alternativeAnswerContainer.apply {
            answerMediator.view.let {
                if (it == null) {
                    if (answerMediator is IButtonListAnswerMediator) {
                        autoDispose {
                            answerMediator.answerButtonPublisher.subscribe { buttonList ->
                                if (buttonList.isEmpty())
                                    AnimationUtils.fadeAnimation(this,
                                            true,
                                            Delay.SHORT.get(questContext)) {
                                        removeAllViews()
                                    }
                                else addView(buttonList.data)
                            }
                        }
                    }
                } else {
                    addView(it)
                }
            }
        }
        val defaultAnswerInputVisible = !alternativeAnswerIsPresent()
        viewHolder.defaultAnswerInputContainer.visibility = if (defaultAnswerInputVisible) VISIBLE else GONE
        val showAnswerButton = !alternativeAnswerWithButtonListIsPresent() && answerButtonVisible()
        if (defaultAnswerInputVisible) {
            viewHolder.defaultAnswerInput.visibility = if (defaultAnswerInputVisible) VISIBLE else GONE
            viewHolder.defaultAnswerButton.visibility = if (showAnswerButton) VISIBLE else GONE
            viewHolder.alternativeAnswerContainer.visibility = if (alternativeAnswerWithButtonListIsPresent()) VISIBLE else GONE
            val answerInput: EditText? = if (alternativeAnswerInputIsPresent()) {
                val contentAnswerInput = contentMediator.answerInput
                if (contentAnswerInput != null && contentAnswerInput.visibility == VISIBLE) {
                    contentAnswerInput.inputType = quest.getAnswerInputType()
                    contentAnswerInput
                } else null
            } else
                viewHolder.defaultAnswerInput
            answerInput?.apply {
                inputType = quest.getAnswerInputType()
                imeOptions = IME_ACTION_DONE
                autoDispose {
                    editorActionEvents().map { it ->
                        it.actionId() == IME_ACTION_DONE
                    }.subscribe { if (it) answerFunction() }
                }
            }
        } else if (showAnswerButton) {
            (viewHolder.defaultAnswerButton.layoutParams as LinearLayout.LayoutParams).apply {
                weight = 1f
                viewHolder.defaultAnswerButton.layoutParams = this
            }
        }
        updateVisibilityOfViews(false)
        if (viewHolder.defaultAnswerButton.visibility == VISIBLE)
            autoDispose {
                viewHolder.defaultAnswerButton.throttleClicks {
                    answerFunction()
                }
            }
    }

    private fun answerFunction() {
        if (!alternativeAnswerWithButtonListIsPresent()) {
            answerCallback.onNext((if (alternativeAnswerInputIsPresent())
                contentMediator.answerInput
            else
                viewHolder.defaultAnswerInput)?.text.toString())
        }
    }

    override fun onQuestAttach(rootContentContainer: ViewGroup) {
        titleMediator.onQuestAttach(viewHolder.titleContainer)
        contentMediator.onQuestAttach(viewHolder.contentContainer)
        answerMediator.onQuestAttach(viewHolder.alternativeAnswerContainer)
        viewHolder.view.addOnLayoutChangeListener(this)
    }

    override fun onQuestStart(delayedPlay: Boolean) {
        viewHolder.titleContainer.requestLayout()
        viewHolder.contentContainer.requestLayout()
        viewHolder.alternativeAnswerContainer.requestLayout()
        titleMediator.onQuestStart(delayedPlay)
        contentMediator.onQuestStart(delayedPlay)
        answerMediator.onQuestStart(delayedPlay)
        requestFocus()
        questHasState(DELAYED_PLAY) {
            updateVisibilityOfViews(!it)
        }
    }

    override fun onQuestPlay(delayedPlay: Boolean) {
        updateVisibilityOfViews(true)
        titleMediator.onQuestPlay(delayedPlay)
        contentMediator.onQuestPlay(delayedPlay)
        if (delayedPlay) {
            if (alternativeAnswerWithButtonListIsPresent()) {
                answerMediator.onQuestPlay(true)
            } else {
                val duration = questContext.questDelayedPlayAnimationDuration
                viewHolder.answerContainer.let {
                    it.alpha = 0f
                    it.animate().withLayer().alpha(1f).duration = duration
                }

            }
        } else {
            answerMediator.onQuestPlay(false)
        }
        requestFocus()
    }

    override fun onAnswer(answerType: AnswerType): Boolean {
        var showRightAnswerWindow = true
        showRightAnswerWindow = showRightAnswerWindow && titleMediator.onAnswer(answerType)
        showRightAnswerWindow = showRightAnswerWindow && contentMediator.onAnswer(answerType)
        showRightAnswerWindow = showRightAnswerWindow && answerMediator.onAnswer(answerType)
        return showRightAnswerWindow
    }

    override fun onQuestReplay() {
        titleMediator.onQuestReplay()
        contentMediator.onQuestReplay()
        answerMediator.onQuestReplay()
        requestFocus()
    }

    override fun onQuestPause() {
        titleMediator.onQuestPause()
        contentMediator.onQuestPause()
        answerMediator.onQuestPause()
    }

    override fun onQuestResume() {
        titleMediator.onQuestResume()
        contentMediator.onQuestResume()
        answerMediator.onQuestResume()
        requestFocus()
    }

    override fun onQuestStop() {
        titleMediator.onQuestStop()
        contentMediator.onQuestStop()
        answerMediator.onQuestStop()
    }

    private fun alternativeAnswerInputIsPresent(): Boolean {
        val answerInput = contentMediator.answerInput
        return answerInput != null && answerInput.visibility == VISIBLE
    }

    private fun alternativeAnswerWithButtonListIsPresent(): Boolean {
        if (answerMediator is IButtonListAnswerMediator) {
            val buttonList = answerMediator.answerButtonPublisher.values
            val answerView = answerMediator.view
            return buttonList.isNotEmpty() ||
                    answerView != null && answerView.visibility == VISIBLE
        }
        return false
    }

    private fun alternativeAnswerIsPresent(): Boolean = alternativeAnswerWithButtonListIsPresent()
            || alternativeAnswerInputIsPresent()
            || answerMediator is IAlternativeAnswerMediator

    private fun setErrorColorAndReturnToNormal() {
        val duration = Delay.SHORT.get(questContext).toInt()
        autoDispose {
            Completable.fromRunnable {
                AnimationUtils.getColorAnimator(questContext, R.color.green,
                        R.color.red,
                        duration,
                        viewHolder.defaultAnswerInputContainer, BounceInterpolator()
                ).start()
            }.subscribeOn(questContext.schedulerProvider.ui())
                    .delay(Delay.LONG.get(questContext),
                            TimeUnit.MILLISECONDS, questContext.schedulerProvider.ui()).concatWith(
                    Completable.fromRunnable {
                        AnimationUtils.getColorAnimator(questContext, R.color.red,
                                R.color.green,
                                duration,
                                viewHolder.defaultAnswerInputContainer
                        ).start()
                    }).subscribe()
        }
    }

    override fun deactivate() {
        view.removeOnLayoutChangeListener(this)
        viewHolder.defaultAnswerInput.setOnEditorActionListener(null)
        onQuestStop()
        titleMediator.deactivate()
        contentMediator.deactivate()
        answerMediator.deactivate()
        dispose()
    }

    override fun detachView() {
        titleMediator.detachView()
        contentMediator.detachView()
        answerMediator.detachView()
    }

    private fun requestFocus() = questHasState(DELAYED_PLAY) { delayed ->
        questHasState(PLAYED) { played ->
            if (delayed && !played)
                hideKeyboard(viewHolder.defaultAnswerInput)
            else {
                if (alternativeAnswerWithButtonListIsPresent())
                    hideKeyboard(viewHolder.defaultAnswerInput)
                if (contentMediator.answerInput == null) {
                    if (viewHolder.defaultAnswerInput.visibility == VISIBLE)
                        showKeyboard(viewHolder.defaultAnswerInput)
                    else
                        hideKeyboard(viewHolder.defaultAnswerInput)
                } else if (contentMediator.answerInput?.visibility == VISIBLE) {
                    showKeyboard(contentMediator.answerInput!!)
                }
            }
        }
    }

    private fun showKeyboard(view: View) {
        KeyboardHost.showKeyboard(questContext, view, Delay.KEYBOARD.get(questContext))
    }

    private fun hideKeyboard(view: View) {
        KeyboardHost.hideKeyboard(questContext, view, Delay.KEYBOARD.get(questContext))
    }

    private fun updateVisibilityOfViews(value: Boolean) {
        val contentView = contentMediator.view
        if (contentView == null) {
            viewHolder.contentContainer.visibility = GONE
        } else {
            viewHolder.contentContainer.visibility = if (value) VISIBLE else INVISIBLE
        }
        if (alternativeAnswerWithButtonListIsPresent()) {
            viewHolder.alternativeAnswerContainer.visibility = if (value) VISIBLE else INVISIBLE
        } else {
            viewHolder.answerContainer.visibility = if (value) VISIBLE else INVISIBLE
        }
    }

    override fun onLayoutChange(view: View, left: Int, top: Int, right: Int, bottom: Int, oldLeft: Int,
                                oldTop: Int, oldRight: Int, oldBottom: Int) {
        updateSize()
    }

    internal class QuestViewHolder(context: Context) : ViewHolder(context, R.layout.layout_quest_view) {

        val rootContainer: ViewGroup = view.findViewById(R.id.container_root) as ViewGroup
        val titleContainer: ViewGroup = view.findViewById(R.id.container_title) as ViewGroup
        val answerContainer: ViewGroup = view.findViewById(R.id.container_answer) as ViewGroup
        val contentContainer: ViewGroup = view.findViewById(R.id.container_content) as ViewGroup
        val alternativeAnswerContainer: ViewGroup = view.findViewById(R.id.container_answer_alternative) as ViewGroup
        val defaultAnswerInputContainer: ViewGroup = view.findViewById(R.id.container_answer_default) as ViewGroup
        val defaultAnswerButton: View = view.findViewById(R.id.btn_answer_default)
        val defaultAnswerInput: EditText = view.findViewById(R.id.input_answer_default) as EditText

    }
}